### PR TITLE
TELLIE: Allow run files to be loaded through SNOPController

### PR DESCRIPTION
**Pull request by EdLeming**
This PR allows operators to use the SNOPController to request standardised TELLIE flash sequences down multiple fibres. It also updates the ELLIE gui to allow users to produce said standard run types, and push them to couch.

### CODE REVIEW

File: Source/Experiments/SNOP/SNOPController.m [hunk 1]
```diff
@@ -1017,93 +1057,93 @@ -(IBAction)loadSmellieRunAction:(id)sender
...
             if( [thisRunFile isEqualToString:requestedRunFile]){
                 
+                NSArray*  objs = [[(ORAppDelegate*)[NSApp delegate] document] collectObjectsOfClass:NSClassFromString(@"ELLIEModel")];
+                if (![objs count]) {
+                    NSString* reasonStr = @"ELLIE model not available, add an ELLIE model to your experiment";
+                    NSException* e = [NSException
+                                      exceptionWithName:@"NoEllieModel"
+                                      reason:reasonStr
+                                      userInfo:nil];
+                    [e raise];
```
Comment (tlatorre-uchicago) [inline Source/Experiments/SNOP/SNOPController.m]: Is this going to be caught anywhere? If not, the exception will cause Orca to crash and no one will see the error message.

File: Source/Experiments/SNOP/SNOPController.m [hunk 2]
```diff
@@ -1209,17 +1256,119 @@ - (IBAction) emergencySmellieStopAction:(id)sender
...
+                int no_nodes = [nodes count];
+                int rate = [[tellieRunFile objectForKey:@"trigger_rate"] intValue];
+                float total_time = (no_nodes*no_pulses)/(rate*60)*1.1;
+                [loadedTellieRunTimeLabel setStringValue:[NSString stringWithFormat:@"%0.1f",total_time]];
+            }
+        }
+        //Activate run buttons
+        [tellieStartRunButton setEnabled:YES];
+        [tellieStopRunButton setEnabled:NO];
+        //[tellieEmergencyStop setEnabled:NO];
```
Comment (tlatorre-uchicago) [inline Source/Experiments/SNOP/SNOPController.m]: delete?

File: Source/Experiments/SNOP/SNOPController.m [hunk 3]
```diff
@@ -1209,17 +1256,119 @@ - (IBAction) emergencySmellieStopAction:(id)sender
...
+    // Get ellie model and launch a fire
+    // sequence
+    NSArray*  objs = [[(ORAppDelegate*)[NSApp delegate] document] collectObjectsOfClass:NSClassFromString(@"ELLIEModel")];
+    if (![objs count]) {
+        NSString* reasonStr = @"ELLIE model not available, add an ELLIE model to your experiment";
+        NSException* e = [NSException
+                          exceptionWithName:@"NoEllieModel"
+                          reason:reasonStr
+                          userInfo:nil];
+        [e raise];
```
Comment (tlatorre-uchicago) [inline Source/Experiments/SNOP/SNOPController.m]: same as above. Is this caught anywhere?

**Comment by tlatorre-uchicago** [comment]
It seems like a lot of functionality is in the SNO+ controller. In general, the controller should really only be used to call model functions and do GUI updates. One problem for example, is that exceptions raised in the controller won't be caught anywhere and will cause Orca to crash.

Is it possible to move all of the heavy duty lifting to model functions?

File: Source/Experiments/SNOP/SNOPController.m (hunk 1 above)
Comment (EdLeming) [inline Source/Experiments/SNOP/SNOPController.m]: I have try catches around all calls to server functions. I was expecting the servers to throw exceptions when something is awry. This now just mimics the exception I was expecting the client to throw.

**Comment by EdLeming** [comment]
Hmmm, yeah. This is one of those snowball moments, it was only ever meant to be a couple of lines in the controller, but things grow as the functionality evolves. I'll re-arrange (likely tomorrow) and push up.

File: Source/Experiments/SNOP/SNOPController.m (hunk 2 above)
Comment (EdLeming) [inline Source/Experiments/SNOP/SNOPController.m]: I'm waiting for the TUBii PR to go through to give this stop button a back end. I can delete for now.

**Comment by EdLeming** [comment]
@tlatorre-uchicago I've moved the timing calculations out of the snopcontroller and into the ELLIEModel. The SNOPController now just:

1) Updates gui elements.
2) Performs some key checks on current run conditions.
3) Calls SNOP/ELLIE model functions. 
4) Hosts the TELLIE / SMELLIE run sequence threads

On the 4th point. We could move this to the SNOPModel, but I feel it would make the code slightly more opaque. Do you have a preference?

File: Source/Experiments/SNOP/SNOPModel.m [hunk 4]
```diff
@@ -1758,6 +1764,45 @@ -(void) parseSmellieRunFileDocs:(id)aResult
     [[NSNotificationCenter defaultCenter] postNotificationName: @"SmellieRunFilesLoaded" object:nil];
 }
 
+- (void) getTellieRunFiles
+{
+    //Set SmellieRunFiles to nil
```
Comment (tlatorre-uchicago) [inline Source/Experiments/SNOP/SNOPModel.m]: Should be Tellie?

**Comment by tlatorre-uchicago** [comment]
Yes, they should be moved to a model, becaues there can be multiple controllers for a single model when you duplicate a window. I'm not sure whether the SNOPModel or the ELLIE models are the better place to put them, but whatever makes things easier.

**Comment by EdLeming** [comment]
@tlatorre-uchicago, I've moved the thread to the ELLIEModel. Local testing has been a success.

**Comment by tlatorre-uchicago** [comment]
ok, if you can just squash to a single commit with a messge like:

```
SNO+: [main message]

[details]
```

I'll merge.

**Comment by EdLeming** [comment]
👍  I'll get this pushed up in the morning

**Comment by EdLeming** [comment]
@tlatorre-uchicago. This guy should be ready to go.

**Comment by tlatorre-uchicago** [comment]
Could you remove the empty whitespace from the commit?

**Comment by tlatorre-uchicago** [comment]
merged.

**Pull request closed without merging**